src/components/item/item-reorder-gesture.ts

```diff
@@ -0,0 +1,137 @@
+import {Item} from './item';
+import {List} from '../list/list';
+import {closest, Coordinates, pointerCoord, CSS, raf} from '../../util/dom';
+
+/**
+ * @private
+ */
+export class ItemReorderGesture {
+  private selectedItem: Item = null;
+  private offset: Coordinates;
+  private itemHeight: number;
+  private lastToIndex: number;
+  private emptyZone: boolean;
+
+  private onTouchstart: Function;
+  private onTouchmove: Function;
+  private onTouchend: Function;
+
+  constructor(public list: List) {
+    this.listen();
+  }
+
+  private onDragStart(ev: any) {
+    let itemEle = ev.target;
+    if (itemEle.nodeName !== 'ION-REORDER') {
+      return;
+    }
+
+    let item = itemEle['$ionComponent'];
+    if (!item) {
+      console.error('item does not contain ion component');
+      return;
+    }
+    ev.preventDefault();
+
+    // Preparing state    
+    this.offset = pointerCoord(ev);
+    this.selectedItem = item;
+    this.itemHeight = item.height();
+    this.lastToIndex = item.index;
+    item.setCssClass('reorder-active', true);
+  }
+
+  private onDrag(ev: any) {
+    if (!this.selectedItem) {
+      return;
+    }
+    ev.preventDefault();
+
+    // Update selected item position    
+    let coord = pointerCoord(ev);
+    let ydiff = Math.round(coord.y - this.offset.y);
+    let style = 'translateY(' + ydiff + 'px)';
+    this.selectedItem.setCssStyle(CSS.transform, style);
+
+    // Hit test    
+    let item = this.itemForCoord(coord);
+    if (!item) {
+      this.emptyZone = true;
+      return;
+    }
+
+    // Move surrounding items if needed 
+    let fromIndex = this.selectedItem.index;    
+    let toIndex = item.index;
+    if (toIndex !== this.lastToIndex || this.emptyZone) {
+      this.lastToIndex = toIndex;
+      this.emptyZone = false;
+      this.list.reorderMove(fromIndex, toIndex, this.itemHeight);
+    }
+  }
+ 
+  private onDragEnd(ev: any) {
+    if (!this.selectedItem) {
+      return;
+    }
+
+    let toIndex = this.lastToIndex;
+    let fromIndex = this.selectedItem.index;
+
+    this.selectedItem.setCssClass('reorder-active', false);
+    this.selectedItem = null;
+    this.list.reorderEmit(fromIndex, toIndex);
+  }
+
+  private itemForCoord(coord: Coordinates): Item {
+    let element = <any>document.elementFromPoint(this.offset.x - 100, coord.y);
+    if (!element) {
+      return null;
+    }
+    element = closest(element, 'ion-item', true);
+    if (!element) {
+      return null;
+    }
+    let item = <Item>(<any>element)['$ionComponent'];
+    if (!item) {
+      return null;
+    }
+    return item;
+  }
+
+  /**
+   * @private
+   */
+  listen() {
+    let ele = this.list.getNativeElement();
+    this.onTouchstart = (ev: any) => this.onDragStart(ev);
```

Comment:
Why does this gesture provide it's own listen() and unlisten(), and manually add/remove touch events? Also, what about mouse events?